File: loom-compiler/src/tokens/Token.cpp
//
// Created by Wylan Shoemaker on 8/6/25.
//

#include "Token.hpp"
#include <sstream>
#include <iomanip>
#include <utility>
#include <type_traits>

namespace lang {
    // --- Token ---

    Token::Token(const TokenType type,
                 const std::string_view lexeme,
                 TokenSpan span,
                 TokenValue value,
                 TokenCategory category)
        : type_(type),
          text_(lexeme),
          span_(std::move(span)),
          value_(std::move(value)),
          category_(category) {
        // Auto-classify when not explicitly provided (except truly invalid tokens)
        if (category_ == TokenCategory::Error && type_ != TokenType::Invalid) {
            category_ = classifyTokenType(type_);
        }
    }

    TokenType Token::type() const noexcept {
        return type_;
    }

    const std::string &Token::text() const noexcept {
        return text_;
    }

    const TokenSpan &Token::span() const noexcept {
        return span_;
    }

    TokenCategory Token::category() const noexcept {
        return category_;
    }

    const TokenValue &Token::value() const noexcept {
        return value_;
    }

    bool Token::hasValue() const noexcept {
        return !std::holds_alternative<std::monostate>(value_);
    }

    bool Token::is(TokenType t) const noexcept {
        return type_ == t;
    }

    bool Token::isCategory(TokenCategory c) const noexcept {
        return category_ == c;
    }

    std::string Token::to_string() const {
        std::ostringstream oss;
        oss << "Token(type:" << tokenTypeName(type_)
                << ", text:\"" << text_ << "\""
                << ", span:" << span_.to_string()
                << ", category:" << tokenCategoryName(category_);

        std::visit([&oss]<typename T>(T &&val) {
            using Decayed = std::decay_t<T>;
            if constexpr (std::is_same_v<Decayed, std::monostate>) {
                oss << ", value: none";
            } else if constexpr (std::is_same_v<Decayed, std::string>) {
                oss << ", value: \"" << val << "\"";
            } else if constexpr (std::is_same_v<Decayed, int64_t>) {
                oss << ", value: " << val;
            } else if constexpr (std::is_same_v<Decayed, double>) {
                oss << ", value: " << std::fixed << std::setprecision(6) << val;
            } else if constexpr (std::is_same_v<Decayed, bool>) {
                oss << ", value: " << (val ? "true" : "false");
            } else if constexpr (std::is_same_v<Decayed, NumericLiteral>) {
                oss << ", value: NumericLiteral{digits=\"" << val.digits
                    << "\", base=" << val.base
                    << ", isFloat=" << (val.isFloat ? "true" : "false")
                    << ", suffix=\"" << val.suffix << "\"}";
            }
        }, value_);

        oss << ")";
        return oss.str();
    }

    bool Token::operator==(const Token &other) const noexcept {
        return type_ == other.type_ &&
               text_ == other.text_ &&
               span_ == other.span_ &&
               value_ == other.value_ &&
               category_ == other.category_;
    }

    bool Token::operator!=(const Token &other) const noexcept {
        return !(*this == other);
    }
} // namespace lang

// --- Hash support for unordered_map / unordered_set ---

namespace std {
    std::size_t hash<lang::Token>::operator()(const lang::Token &tok) const noexcept {
        // 64-bit FNV-1a for stable ordering across runs; fold to size_t
        constexpr uint64_t FNV_OFFSET = 1469598103934665603ull;
        constexpr uint64_t FNV_PRIME = 1099511628211ull;

        auto mix64 = [&](uint64_t h, uint64_t v) noexcept -> uint64_t {
            h ^= v;
            h *= FNV_PRIME;
            return h;
        };

        uint64_t h = FNV_OFFSET;

        // type
        h = mix64(h, static_cast<uint64_t>(tok.type()));

        // category
        h = mix64(h, static_cast<uint64_t>(tok.category()));

        // text
        const std::string &s = tok.text();
        for (unsigned char c: s) {
            h = mix64(h, c);
        }

        // span
        const auto &sp = tok.span();
        // include file path to reduce collisions
        for (unsigned char c: sp.file) {
            h = mix64(h, c);
        }
        h = mix64(h, sp.startLine);
        h = mix64(h, sp.startColumn);
        h = mix64(h, sp.endLine);
        h = mix64(h, sp.endColumn);

        if constexpr (sizeof(std::size_t) >= sizeof(uint64_t)) {
            return h;
        } else {
            return static_cast<std::size_t>(
                (h >> (8 * sizeof(std::size_t))) ^ (h & ((1ull << (8 * sizeof(std::size_t))) - 1))
            );
        }
    }
}

// --- Token names and classification ---

namespace lang {
    TokenCategory classifyTokenType(TokenType type) noexcept {
        using TT = TokenType;
        switch (type) {
            // Literals
            case TT::Char:
            case TT::False:
            case TT::Null:
            case TT::Number:
            case TT::String:
            case TT::True:
                return TokenCategory::Literal;
            // Keywords
            case TT::As:
            case TT::Atomic:
            case TT::Bit:
            case TT::Bool:
            case TT::Break:
            case TT::Builder:
            case TT::Case:
            case TT::Class:
            case TT::Const:
            case TT::Continue:
            case TT::Default:
            case TT::Do:
            case TT::Elif:
            case TT::Else:
            case TT::Enum:
            case TT::Fin:
            case TT::For:
            case TT::Func:
            case TT::If:
            case TT::Import:
            case TT::In:
            case TT::Internal:
            case TT::Let:
            case TT::Loop:
            case TT::Mod:
            case TT::New:
            case TT::Priv:
            case TT::Prot:
            case TT::Pub:
            case TT::Ret:
            case TT::Rev:
            case TT::Self:
            case TT::Step:
            case TT::Struct:
            case TT::Super:
            case TT::Switch:
            case TT::This:
            case TT::Var:
            case TT::While:
                return TokenCategory::Keyword;
            // Built-in types are considered keywords or identifiers depending on language design; choose Keyword here.
            case TT::Byte:
            case TT::f16:
            case TT::f32:
            case TT::f64:
            case TT::i8:
            case TT::i16:
            case TT::i32:
            case TT::i64:
            case TT::u8:
            case TT::u16:
            case TT::u32:
            case TT::u64:
                return TokenCategory::Keyword;
            // Operators
            case TT::And:
            case TT::Arrow:
            case TT::DoubleEqual:
            case TT::Equal:
            case TT::Greater:
            case TT::GreaterEqual:
            case TT::Less:
            case TT::LessEqual:
            case TT::Minus:
            case TT::Not:
            case TT::NotEqual:
            case TT::Or:
            case TT::Percent:
            case TT::Plus:
            case TT::Range:
            case TT::Range_Inclusive:
            case TT::Slash:
            case TT::Star:
                return TokenCategory::Operator;
            // Punctuation
            case TT::Colon:
            case TT::Comma:
            case TT::DoubleColon:
            case TT::Dot:
            case TT::LBrace:
            case TT::LBracket:
            case TT::LParen:
            case TT::Question:
            case TT::RBrace:
            case TT::RBracket:
            case TT::RParen:
            case TT::Semicolon:
                return TokenCategory::Punctuation;
            // Identifier token kind stands alone
            case TT::Identifier:
                return TokenCategory::Identifier;
            // Special
            case TT::EndOfFile:
                return TokenCategory::Eof;
            case TT::Invalid:
                return TokenCategory::Error;
        }
        return TokenCategory::Error;
    }

    std::string_view tokenTypeName(TokenType type) noexcept {
        switch (type) {
#define CASE_NAME(x) case TokenType::x: return #x;
            CASE_NAME(Char)
            CASE_NAME(False)
            CASE_NAME(Identifier)
            CASE_NAME(Null)
            CASE_NAME(Number)
            CASE_NAME(String)
            CASE_NAME(True)
            CASE_NAME(As)
            CASE_NAME(Atomic)
            CASE_NAME(Bit)
            CASE_NAME(Bool)
            CASE_NAME(Break)
            CASE_NAME(Builder)
            CASE_NAME(Case)
            CASE_NAME(Class)
            CASE_NAME(Const)
            CASE_NAME(Continue)
            CASE_NAME(Default)
            CASE_NAME(Do)
            CASE_NAME(Elif)
            CASE_NAME(Else)
            CASE_NAME(Enum)
            CASE_NAME(Fin)
            CASE_NAME(For)
            CASE_NAME(Func)
            CASE_NAME(If)
            CASE_NAME(Import)
            CASE_NAME(In)
            CASE_NAME(Internal)
            CASE_NAME(Let)
            CASE_NAME(Loop)
            CASE_NAME(Mod)
            CASE_NAME(New)
            CASE_NAME(Priv)
            CASE_NAME(Prot)
            CASE_NAME(Pub)
            CASE_NAME(Ret)
            CASE_NAME(Rev)
            CASE_NAME(Self)
            CASE_NAME(Step)
            CASE_NAME(Struct)
            CASE_NAME(Super)
            CASE_NAME(Switch)
            CASE_NAME(This)
            CASE_NAME(Var)
            CASE_NAME(While)
            CASE_NAME(Byte)
            CASE_NAME(f16)
            CASE_NAME(f32)
            CASE_NAME(f64)
            CASE_NAME(i8)
            CASE_NAME(i16)
            CASE_NAME(i32)
            CASE_NAME(i64)
            CASE_NAME(u8)
            CASE_NAME(u16)
            CASE_NAME(u32)
            CASE_NAME(u64)
            CASE_NAME(And)
            CASE_NAME(Arrow)
            CASE_NAME(DoubleEqual)
            CASE_NAME(Equal)
            CASE_NAME(Greater)
            CASE_NAME(GreaterEqual)
            CASE_NAME(Less)
            CASE_NAME(LessEqual)
            CASE_NAME(Minus)
            CASE_NAME(Not)
            CASE_NAME(NotEqual)
            CASE_NAME(Or)
            CASE_NAME(Percent)
            CASE_NAME(Plus)
            CASE_NAME(Range)
            CASE_NAME(Range_Inclusive)
            CASE_NAME(Slash)
            CASE_NAME(Star)
            CASE_NAME(Colon)
            CASE_NAME(Comma)
            CASE_NAME(DoubleColon)
            CASE_NAME(Dot)
            CASE_NAME(LBrace)
            CASE_NAME(LBracket)
            CASE_NAME(LParen)
            CASE_NAME(Question)
            CASE_NAME(RBrace)
            CASE_NAME(RBracket)
            CASE_NAME(RParen)
            CASE_NAME(Semicolon)
            CASE_NAME(EndOfFile)
            CASE_NAME(Invalid)
#undef CASE_NAME
        }
        return "Unknown";
    }

    std::string_view tokenCategoryName(TokenCategory category) noexcept {
        switch (category) {
            case TokenCategory::Literal: return "Literal";
            case TokenCategory::Keyword: return "Keyword";
            case TokenCategory::Operator: return "Operator";
            case TokenCategory::Punctuation: return "Punctuation";
            case TokenCategory::Identifier: return "Identifier";
            case TokenCategory::Whitespace: return "Whitespace";
            case TokenCategory::Comment: return "Comment";
            case TokenCategory::Error: return "Error";
            case TokenCategory::Eof: return "Eof";
        }
        return "Unknown";
    }
} // namespace lang

File: compiler/src/tokens/Token.go
package tokens

import (
	"fmt"
)

// TokenType mirrors the C++ loom::TokenType
type TokenType uint16

const (
	// Literals
	TokenChar TokenType = iota
	TokenFalse
	TokenIdentifier
	TokenNull
	TokenNumber
	TokenString
	TokenTrue

	// Keywords
	TokenAlias
	TokenAs
	TokenAtomic
	TokenBit
	TokenBool
	TokenBreak
	TokenBuilder
	TokenCase
	TokenClass
	TokenConst
	TokenContinue
	TokenDefault
	TokenDo
	TokenElif
	TokenElse
	TokenEnum
	TokenFin
	TokenFor
	TokenFunc
	TokenIf
	TokenImport
	TokenIn
	TokenInternal
	TokenLet
	TokenLoop
	TokenMod
	TokenNew
	TokenPriv
	TokenProt
	TokenPub
	TokenRet
	TokenRev
	TokenSelf
	TokenStep
	TokenStruct
	TokenSuper
	TokenSwitch
	TokenThis
	TokenVar
	TokenWhile

	// Built-in Types
	TokenByte
	TokenF16
	TokenF32
	TokenF64
	TokenI8
	TokenI16
	TokenI32
	TokenI64
	TokenU8
	TokenU16
	TokenU32
	TokenU64

	// Operators
	TokenAnd
	TokenArrow
	TokenDoubleEqual
	TokenEqual
	TokenGreater
	TokenGreaterEqual
	TokenLess
	TokenLessEqual
	TokenMinus
	TokenNot
	TokenNotEqual
	TokenOr
	TokenPercent
	TokenPlus
	TokenRange
	TokenRangeInclusive
	TokenSlash
	TokenStar

	// Symbols
	TokenColon
	TokenComma
	TokenDoubleColon
	TokenDot
	TokenLBrace
	TokenLBracket
	TokenLParen
	TokenQuestion
	TokenRBrace
	TokenRBracket
	TokenRParen
	TokenSemicolon

	// Special
	TokenEndOfFile
	TokenInvalid
)

// TokenCategory mirrors loom::TokenCategory
type TokenCategory uint8

const (
	CategoryLiteral TokenCategory = iota
	CategoryKeyword
	CategoryOperator
	CategoryPunctuation
	CategoryIdentifier
	CategoryWhitespace
	CategoryComment
	CategoryError
	CategoryEof
)

// NumericLiteral preserves numeric metadata
type NumericLiteral struct {
	Digits  string
	Base    int
	IsFloat bool
	Suffix  string
}

// TokenValue value uses a tagged union style via Go interface{} with helpers
type TokenValue interface{}

// Token mirrors loom::Token
type Token struct {
	Type     TokenType
	Text     string
	Span     TokenSpan
	Value    TokenValue
	Category TokenCategory
}

// NewToken constructs a token; if category is CategoryError but type != TokenInvalid, auto-classify
func NewToken(typ TokenType, lexeme string, span TokenSpan, value TokenValue, category TokenCategory) Token {
	if category == CategoryError && typ != TokenInvalid {
		category = ClassifyTokenType(typ)
	}
	return Token{Type: typ, Text: lexeme, Span: span, Value: value, Category: category}
}

func (t Token) Is(typ TokenType) bool             { return t.Type == typ }
func (t Token) IsCategory(cat TokenCategory) bool { return t.Category == cat }

func (t Token) String() string {
	return fmt.Sprintf("Token {\n\ttype: %s \n\ttext: \"%s\" \n\tspan: %s \n\tcategory: %s \n\tvalue: %s\n}",
		TokenTypeName(t.Type), t.Text, t.Span.String(), TokenCategoryName(t.Category), tokenValueString(t.Value))
}

func tokenValueString(v TokenValue) string {
	switch val := v.(type) {
	case nil:
		return "none"
	case string:
		return fmt.Sprintf("\"%s\"", val)
	case int, int8, int16, int32, int64:
		return fmt.Sprintf("%d", val)
	case uint, uint8, uint16, uint32, uint64:
		return fmt.Sprintf("%d", val)
	case float32, float64:
		return fmt.Sprintf("%g", val)
	case bool:
		if val {
			return "true"
		}
		return "false"
	case NumericLiteral:
		return fmt.Sprintf("NumericLiteral{digits=\"%s\", base=%d, isFloat=%t, suffix=\"%s\"}", val.Digits, val.Base, val.IsFloat, val.Suffix)
	default:
		return fmt.Sprintf("%v", val)
	}
}

// ClassifyTokenType maps a TokenType to its category
func ClassifyTokenType(t TokenType) TokenCategory {
	switch t {
	// Literals
	case TokenChar, TokenFalse, TokenNull, TokenNumber, TokenString, TokenTrue:
		return CategoryLiteral
	// Keywords
	case TokenAlias, TokenAs, TokenAtomic, TokenBit, TokenBool, TokenBreak, TokenBuilder, TokenCase, TokenClass, TokenConst, TokenContinue, TokenDefault, TokenDo, TokenElif, TokenElse, TokenEnum, TokenFin, TokenFor, TokenFunc, TokenIf, TokenImport, TokenIn, TokenInternal, TokenLet, TokenLoop, TokenMod, TokenNew, TokenPriv, TokenProt, TokenPub, TokenRet, TokenRev, TokenSelf, TokenStep, TokenStruct, TokenSuper, TokenSwitch, TokenThis, TokenVar, TokenWhile:
		return CategoryKeyword
	// Built-in types -> treat as keywords
	case TokenByte, TokenF16, TokenF32, TokenF64, TokenI8, TokenI16, TokenI32, TokenI64, TokenU8, TokenU16, TokenU32, TokenU64:
		return CategoryKeyword
	// Operators
	case TokenAnd, TokenArrow, TokenDoubleEqual, TokenEqual, TokenGreater, TokenGreaterEqual, TokenLess, TokenLessEqual, TokenMinus, TokenNot, TokenNotEqual, TokenOr, TokenPercent, TokenPlus, TokenRange, TokenRangeInclusive, TokenSlash, TokenStar:
		return CategoryOperator
	// Punctuation
	case TokenColon, TokenComma, TokenDoubleColon, TokenDot, TokenLBrace, TokenLBracket, TokenLParen, TokenQuestion, TokenRBrace, TokenRBracket, TokenRParen, TokenSemicolon:
		return CategoryPunctuation
	case TokenIdentifier:
		return CategoryIdentifier
	case TokenEndOfFile:
		return CategoryEof
	case TokenInvalid:
		return CategoryError
	}
	return CategoryError
}

// TokenTypeName returns a stable string name for a token type
func TokenTypeName(t TokenType) string {
	switch t {
	case TokenChar:
		return "Char"
	case TokenFalse:
		return "False"
	case TokenIdentifier:
		return "Identifier"
	case TokenNull:
		return "Null"
	case TokenNumber:
		return "Number"
	case TokenString:
		return "String"
	case TokenTrue:
		return "True"
	case TokenAlias:
		return "Alias"
	case TokenAs:
		return "As"
	case TokenAtomic:
		return "Atomic"
	case TokenBit:
		return "Bit"
	case TokenBool:
		return "Bool"
	case TokenBreak:
		return "Break"
	case TokenBuilder:
		return "Builder"
	case TokenCase:
		return "Case"
	case TokenClass:
		return "Class"
	case TokenConst:
		return "Const"
	case TokenContinue:
		return "Continue"
	case TokenDefault:
		return "Default"
	case TokenDo:
		return "Do"
	case TokenElif:
		return "Elif"
	case TokenElse:
		return "Else"
	case TokenEnum:
		return "Enumerator"
	case TokenFin:
		return "Final"
	case TokenFor:
		return "For"
	case TokenFunc:
		return "Function"
	case TokenIf:
		return "If"
	case TokenImport:
		return "Import"
	case TokenIn:
		return "In"
	case TokenInternal:
		return "Internal"
	case TokenLet:
		return "Let"
	case TokenLoop:
		return "Loop"
	case TokenMod:
		return "Module"
	case TokenNew:
		return "New"
	case TokenPriv:
		return "Private"
	case TokenProt:
		return "Protected"
	case TokenPub:
		return "Public"
	case TokenRet:
		return "Return"
	case TokenRev:
		return "Reverse"
	case TokenSelf:
		return "Self"
	case TokenStep:
		return "Step"
	case TokenStruct:
		return "Structure"
	case TokenSuper:
		return "Super"
	case TokenSwitch:
		return "Switch"
	case TokenThis:
		return "This"
	case TokenVar:
		return "Variable"
	case TokenWhile:
		return "While"
	case TokenByte:
		return "Byte"
	case TokenF16:
		return "f16"
	case TokenF32:
		return "f32"
	case TokenF64:
		return "f64"
	case TokenI8:
		return "i8"
	case TokenI16:
		return "i16"
	case TokenI32:
		return "i32"
	case TokenI64:
		return "i64"
	case TokenU8:
		return "u8"
	case TokenU16:
		return "u16"
	case TokenU32:
		return "u32"
	case TokenU64:
		return "u64"
	case TokenAnd:
		return "And"
	case TokenArrow:
		return "Arrow"
	case TokenDoubleEqual:
		return "DoubleEqual"
	case TokenEqual:
		return "Equal"
	case TokenGreater:
		return "Greater"
	case TokenGreaterEqual:
		return "GreaterEqual"
	case TokenLess:
		return "Less"
	case TokenLessEqual:
		return "LessEqual"
	case TokenMinus:
		return "Minus"
	case TokenNot:
		return "Not"
	case TokenNotEqual:
		return "NotEqual"
	case TokenOr:
		return "Or"
	case TokenPercent:
		return "Percent"
	case TokenPlus:
		return "Plus"
	case TokenRange:
		return "Range"
	case TokenRangeInclusive:
		return "Range Inclusive"
	case TokenSlash:
		return "Slash"
	case TokenStar:
		return "Star"
	case TokenColon:
		return "Colon"
	case TokenComma:
		return "Comma"
	case TokenDoubleColon:
		return "DoubleColon"
	case TokenDot:
		return "Dot"
	case TokenLBrace:
		return "Left Brace"
	case TokenLBracket:
		return "Left Bracket"
	case TokenLParen:
		return "Left Paren"
	case TokenQuestion:
		return "Question"
	case TokenRBrace:
		return "Right Brace"
	case TokenRBracket:
		return "Right Bracket"
	case TokenRParen:
		return "Right Paren"
	case TokenSemicolon:
		return "Semicolon"
	case TokenEndOfFile:
		return "End Of File"
	case TokenInvalid:
		return "Invalid"
	}
	return "Unknown"
}

// TokenCategoryName returns string name for category
func TokenCategoryName(c TokenCategory) string {
	switch c {
	case CategoryLiteral:
		return "Literal"
	case CategoryKeyword:
		return "Keyword"
	case CategoryOperator:
		return "Operator"
	case CategoryPunctuation:
		return "Punctuation"
	case CategoryIdentifier:
		return "Identifier"
	case CategoryWhitespace:
		return "Whitespace"
	case CategoryComment:
		return "Comment"
	case CategoryError:
		return "Error"
	case CategoryEof:
		return "Eof"
	}
	return "Unknown"
}
